src/ServiceControl.Monitoring/Raw/RawDataProvider.cs

```diff
@@ -1,29 +1,83 @@
 ﻿namespace ServiceControl.Monitoring.Raw
 {
+    using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
-    using Http;
+    using System.Linq;
     using Newtonsoft.Json.Linq;
 
-    /// <summary>
-    /// The raw endpoint data provider, consuming data with <see cref="Consume"/> and providing them with raw data.
-    /// </summary>
-    public class RawDataProvider : IEndpointDataProvider, IRawDataConsumer
+    public class RawDataProvider : IRawDataConsumer
     {
-        public const string Name = "Raw";
-
-        /// <summary>
-        /// Consumes a new portion of data.
-        /// </summary>
         public void Consume(IReadOnlyDictionary<string, string> headers, JObject data)
         {
-            var name = headers.GetOriginatingEndpoint();
-            contexts.AddOrUpdate(name, data, (context, currentData) => data);
+            var endpointName = headers.GetOriginatingEndpoint();
+            var timestamp = data["Timestamp"].Value<DateTime>();
+            var criticalTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Critical Time");
+            var criticalTime = criticalTimeMetrics["Histogram"]["Mean"].Value<float>();
+            var processingTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Processing Time");
+            var processingTime = processingTimeMetrics["Histogram"]["Mean"].Value<float>();
+
+            // ordered collections are not threadsafe
+            var endpointLock = endpointLocks.GetOrAdd(endpointName, new object());
+
+            lock (endpointLock)
+            {
+                if (!criticalTimeValues.ContainsKey(endpointLock))
+                    criticalTimeValues[endpointLock] = new SortedList<DateTime, float>();
+                criticalTimeValues[endpointLock].Add(timestamp, criticalTime);
+
+                if (!processingTimeValues.ContainsKey(endpointLock))
+                    processingTimeValues[endpointLock] = new SortedList<DateTime, float>();
+                processingTimeValues[endpointLock].Add(timestamp, processingTime);
+            }
+        }
+
+        public dynamic GetValues(TimeSpan range)
+        {
+            var historicMax = DateTime.UtcNow.Subtract(range);
+            var result = new List<EndpointMetrics>();
+            foreach (var endpoint in endpointLocks)
+            {
+                var metrics = new EndpointMetrics(endpoint.Key);
+
+                var endpointLock = endpoint.Value;
+                lock (endpointLock)
+                {
+                    // check because of potential race condition
+                    if (criticalTimeValues.ContainsKey(endpointLock))
+                    {
+                        var endpointCriticalTime = criticalTimeValues[endpointLock];
+                        metrics.CriticalTime = endpointCriticalTime.SkipWhile(x => x.Key < historicMax).Select(x => x.Value).ToArray();
```

Comment:
in case new data is appended at the end of a sortedList, insert performance is quite fast afaik. As new data in most cases should be inserted at the end (because it has the latest timestamp), this should work fairly well but also handle edge cases (out of order messages) fairly well as the performance on inserting is not that critical (it's just a message handler) while lookup performance while querying might be more important.